# Lesson 1: Reading and Summarizing CSV Data

To jump to the recap, click [here](#recap)

## Intro to Pandas

- `pandas` = Python Data Analysis Library (https://pandas.pydata.org/)
- With `pandas` you can do pretty much everything you would in a spreadsheet, plus a whole lot more!

- Why Python + Pandas for spreadsheet data?
  - Working with huge data files and complex calculations
  - Dealing with messy and missing data
  - Merging data from multiple files
  - Timeseries analysis
  - Automate repetitive tasks
  - Huge variety of fully customized graphs for visualizing data

Import the `pandas` library and give it the nickname `pd`

In [1]:
import pandas as pd

*Note: For learning purposes, we are importing each library as we introduce it. In general, it's good practice to collect all your `import` statements together and put them at the start of the notebook.*

# Reading a CSV file

Let's look at the file `weather_yvr.csv` in the sub-folder `data`

- 24 hours of weather measurements at Vancouver Airport
- View it in Jupyter Lab's CSV viewer

We will read the CSV file into our notebook with the function `pd.read_csv`:
- Try typing `pd.re` and then press `Tab` and select `read_csv` from the auto-complete options
- Our input to the `read_csv` function is the file path (including sub-folder) as a string: `'data/weather_yvr.csv'`

In [2]:
pd.read_csv('data/weather_yvr.csv')

Datetime  Hour of Day     Conditions  Temperature (C)  \
0   2018-05-21 22:00:00           22   Mainly Clear             14.8   
1   2018-05-21 23:00:00           23          Clear             13.5   
2   2018-05-22 00:00:00            0          Clear             13.1   
3   2018-05-22 01:00:00            1          Clear             12.9   
4   2018-05-22 02:00:00            2          Clear             12.2   
5   2018-05-22 03:00:00            3          Clear             12.0   
6   2018-05-22 04:00:00            4          Clear             11.9   
7   2018-05-22 05:00:00            5            Fog             10.4   
8   2018-05-22 06:00:00            6   Mainly Sunny             11.0   
9   2018-05-22 07:00:00            7  Partly Cloudy             13.0   
10  2018-05-22 08:00:00            8   Mainly Sunny             14.8   
11  2018-05-22 09:00:00            9  Partly Cloudy             15.8   
12  2018-05-22 10:00:00           10  Partly Cloudy             15.9   
13  2018-05-22 11:00:00           11  Partly Cloudy             16.1   
14  2018-05-22 12:00:00           12   Mainly Sunny             17.4   
15  2018-05-22 13:00:00           13   Mainly Sunny             18.2   
16  2018-05-22 14:00:00           14   Mainly Sunny             18.8   
17  2018-05-22 15:00:00           15   Mainly Sunny             19.5   
18  2018-05-22 16:00:00           16          Sunny             19.4   
19  2018-05-22 17:00:00           17   Mainly Sunny             20.1   
20  2018-05-22 18:00:00           18          Sunny             19.7   
21  2018-05-22 19:00:00           19          Sunny             19.6   
22  2018-05-22 20:00:00           20          Sunny             18.4   
23  2018-05-22 21:00:00           21          Clear             16.1   

    Relative Humidity (%)  
0                    75.0  
1                    76.0  
2                    77.0  
3                    84.0  
4                    88.0  
5                    87.0  
6                    88.0  
7                    97.0  
8                    91.0  
9                    92.0  
10                   82.0  
11                   78.0  
12                   75.0  
13                   80.0  
14                   77.0  
15                   68.0  
16                   64.0  
17                   62.0  
18                   67.0  
19                   64.0  
20                   59.0  
21                   67.0  
22                   70.0  
23                   84.0

- We've displayed the data, but we can't do anything further with this on-screen display
- We need to store the data in a variable

Store the output of `pd.read_csv` in a variable `weather_yvr`:

In [3]:
weather_yvr = pd.read_csv('data/weather_yvr.csv')

In [4]:
weather_yvr

Datetime  Hour of Day     Conditions  Temperature (C)  \
0   2018-05-21 22:00:00           22   Mainly Clear             14.8   
1   2018-05-21 23:00:00           23          Clear             13.5   
2   2018-05-22 00:00:00            0          Clear             13.1   
3   2018-05-22 01:00:00            1          Clear             12.9   
4   2018-05-22 02:00:00            2          Clear             12.2   
5   2018-05-22 03:00:00            3          Clear             12.0   
6   2018-05-22 04:00:00            4          Clear             11.9   
7   2018-05-22 05:00:00            5            Fog             10.4   
8   2018-05-22 06:00:00            6   Mainly Sunny             11.0   
9   2018-05-22 07:00:00            7  Partly Cloudy             13.0   
10  2018-05-22 08:00:00            8   Mainly Sunny             14.8   
11  2018-05-22 09:00:00            9  Partly Cloudy             15.8   
12  2018-05-22 10:00:00           10  Partly Cloudy             15.9   
13  2018-05-22 11:00:00           11  Partly Cloudy             16.1   
14  2018-05-22 12:00:00           12   Mainly Sunny             17.4   
15  2018-05-22 13:00:00           13   Mainly Sunny             18.2   
16  2018-05-22 14:00:00           14   Mainly Sunny             18.8   
17  2018-05-22 15:00:00           15   Mainly Sunny             19.5   
18  2018-05-22 16:00:00           16          Sunny             19.4   
19  2018-05-22 17:00:00           17   Mainly Sunny             20.1   
20  2018-05-22 18:00:00           18          Sunny             19.7   
21  2018-05-22 19:00:00           19          Sunny             19.6   
22  2018-05-22 20:00:00           20          Sunny             18.4   
23  2018-05-22 21:00:00           21          Clear             16.1   

    Relative Humidity (%)  
0                    75.0  
1                    76.0  
2                    77.0  
3                    84.0  
4                    88.0  
5                    87.0  
6                    88.0  
7                    97.0  
8                    91.0  
9                    92.0  
10                   82.0  
11                   78.0  
12                   75.0  
13                   80.0  
14                   77.0  
15                   68.0  
16                   64.0  
17                   62.0  
18                   67.0  
19                   64.0  
20                   59.0  
21                   67.0  
22                   70.0  
23                   84.0

What type of variable is `weather_yvr`?

In [5]:
type(weather_yvr)

pandas.core.frame.DataFrame

- `weather_yvr` is a **DataFrame**, a data type from the `pandas` library
  - A DataFrame is a 2-dimensional array (like a table in a spreadsheet)

- When we display `weather_yvr`, the integer numbers in bold on the left are the DataFrame's **index**

Using `print` to display `weather_yvr` looks a bit different than the IPython display shown in the previous slide

In [6]:
print(weather_yvr)

               Datetime  Hour of Day     Conditions  Temperature (C)  \
0   2018-05-21 22:00:00           22   Mainly Clear             14.8   
1   2018-05-21 23:00:00           23          Clear             13.5   
2   2018-05-22 00:00:00            0          Clear             13.1   
3   2018-05-22 01:00:00            1          Clear             12.9   
4   2018-05-22 02:00:00            2          Clear             12.2   
5   2018-05-22 03:00:00            3          Clear             12.0   
6   2018-05-22 04:00:00            4          Clear             11.9   
7   2018-05-22 05:00:00            5            Fog             10.4   
8   2018-05-22 06:00:00            6   Mainly Sunny             11.0   
9   2018-05-22 07:00:00            7  Partly Cloudy             13.0   
10  2018-05-22 08:00:00            8   Mainly Sunny             14.8   
11  2018-05-22 09:00:00            9  Partly Cloudy             15.8   
12  2018-05-22 10:00:00           10  Partly Cloudy             

# Data at a Glance

`pandas` provides many ways to quickly and easily summarize your data:
- How many rows and columns are there?
- What are all the column names and what type of data is in each column?

- Numerical data: What is the average and range of the values?
- Text data: What are the unique values and how often does each occur?
- How many values are missing in each column?

DataFrame methods:

In [7]:
weather_yvr.head?

- Type `weather_yvr.` followed by `Tab` to see other methods available for the DataFrame

In [8]:
weather_yvr.head()

Datetime  Hour of Day    Conditions  Temperature (C)  \
0  2018-05-21 22:00:00           22  Mainly Clear             14.8   
1  2018-05-21 23:00:00           23         Clear             13.5   
2  2018-05-22 00:00:00            0         Clear             13.1   
3  2018-05-22 01:00:00            1         Clear             12.9   
4  2018-05-22 02:00:00            2         Clear             12.2   

   Relative Humidity (%)  
0                   75.0  
1                   76.0  
2                   77.0  
3                   84.0  
4                   88.0

In [9]:
weather_yvr.head(3)

Datetime  Hour of Day    Conditions  Temperature (C)  \
0  2018-05-21 22:00:00           22  Mainly Clear             14.8   
1  2018-05-21 23:00:00           23         Clear             13.5   
2  2018-05-22 00:00:00            0         Clear             13.1   

   Relative Humidity (%)  
0                   75.0  
1                   76.0  
2                   77.0

In [10]:
weather_yvr.tail()

Datetime  Hour of Day    Conditions  Temperature (C)  \
19  2018-05-22 17:00:00           17  Mainly Sunny             20.1   
20  2018-05-22 18:00:00           18         Sunny             19.7   
21  2018-05-22 19:00:00           19         Sunny             19.6   
22  2018-05-22 20:00:00           20         Sunny             18.4   
23  2018-05-22 21:00:00           21         Clear             16.1   

    Relative Humidity (%)  
19                   64.0  
20                   59.0  
21                   67.0  
22                   70.0  
23                   84.0

In [11]:
weather_yvr.sample(4)

Datetime  Hour of Day     Conditions  Temperature (C)  \
11  2018-05-22 09:00:00            9  Partly Cloudy             15.8   
19  2018-05-22 17:00:00           17   Mainly Sunny             20.1   
1   2018-05-21 23:00:00           23          Clear             13.5   
3   2018-05-22 01:00:00            1          Clear             12.9   

    Relative Humidity (%)  
11                   78.0  
19                   64.0  
1                    76.0  
3                    84.0

Number of rows and columns:

In [12]:
weather_yvr.shape

(24, 5)

- The DataFrame `weather_yvr` has 24 rows and 5 columns
- The index does not count as a column
- Notice there are no parentheses at the end of `weather_yvr.shape`
- `shape` is an **attribute** of the variable `weather_yvr`

We can save `weather_yvr.shape` as a variable:

In [13]:
shape_yvr = weather_yvr.shape
print(shape_yvr)

(24, 5)


In [14]:
type(shape_yvr)

tuple

A tuple is another data type, similar to a list
- Items are enclosed in `()` instead of `[]`
- Tuples are immutable&mdash;you can't modify individual items inside a tuple

Unlike a list, which can contain items of different types, each column of a DataFrame must contain items of the same type.

We can find out the names and data types of each column from the `dtypes` attribute:

In [15]:
weather_yvr.dtypes

Datetime                  object
Hour of Day                int64
Conditions                object
Temperature (C)          float64
Relative Humidity (%)    float64
dtype: object

- In a `pandas` DataFrame, a column containing text data (or containing a mix of text and numbers) is assigned a `dtype` of `object` and is treated as a column of strings

- `int64` and `float64` are integer and float, respectively
  - The `64` at the end means that they are stored as 64-bit numbers in memory
  - These data types are equivalent to `int` and `float` in Python (`pandas` is a just a bit more explicit in how it names them)

If we just want a list of the column names, we can use the `columns` attribute:

In [16]:
weather_yvr.columns

Index(['Datetime', 'Hour of Day', 'Conditions', 'Temperature (C)',
       'Relative Humidity (%)'],
      dtype='object')

# Simple Summary Statistics

In [17]:
weather_yvr.describe()

Hour of Day  Temperature (C)  Relative Humidity (%)
count    24.000000        24.000000              24.000000
mean     11.500000        15.608333              77.166667
std       7.071068         3.119492              10.470109
min       0.000000        10.400000              59.000000
25%       5.750000        12.975000              67.750000
50%      11.500000        15.850000              77.000000
75%      17.250000        18.500000              84.750000
max      23.000000        20.100000              97.000000

- The `describe` method is a way to quickly summarize the averages, extremes, and variability of each numerical data column
- You can look at each statistic individually with methods such as `mean`, `median`, `min`, `max`, and `std`

In [18]:
weather_yvr.mean()

Hour of Day              11.500000
Temperature (C)          15.608333
Relative Humidity (%)    77.166667
dtype: float64

In [19]:
weather_yvr.max()

Datetime                 2018-05-22 21:00:00
Hour of Day                               23
Conditions                             Sunny
Temperature (C)                         20.1
Relative Humidity (%)                     97
dtype: object

The `max` method includes string data from the 'Datetime' and 'Conditions' columns in its calculations, which probably isn't what we want.

Let's check out the documentation for `max`:

In [20]:
weather_yvr.max?

We can use the **keyword argument** `numeric_only` with a Boolean value of `True` to include only the numeric columns:

In [21]:
weather_yvr.max(numeric_only=True)

Hour of Day              23.0
Temperature (C)          20.1
Relative Humidity (%)    97.0
dtype: float64

Auto-complete works for keyword arguments too! You can start typing `nu` inside `weather_yvr.max()`, then press `Tab` and see what happens.

Try pressing `Shift` and `Tab` together after you type the opening `(` in `weather_yvr.max()`
- A little window pops up showing a shortened version of the documentation, including the list of available keyword arguments

<a id="recap"></a>
# Lesson 1 Recap

### Importing `pandas` Library
```
import pandas as pd
```
- Libraries only need to be imported once in a notebook
- It's good practice to consolidate all your `import` commands together near the start of your notebook

### Reading a CSV File

To read a CSV file and store it as a DataFrame variable:
```
df = pd.read_csv('some_cool_data.csv')
```

### Quick and Easy Summaries of a DataFrame

Number of rows and columns (rows first, columns second): 
```
df.shape
```

Names and data types of each column: 
```
df.dtypes
```
Just the names of each column:
```
df.columns
```

#### Rows at a Glance

- First 5 rows:
```
df.head()
```
- Last 5 rows:
```
df.tail()
```
- A random sampling (1 row):
```
df.sample()
```
- The number of rows can be specified as an input to any of the above methods (e.g. `df.tail(7)` returns the last 7 rows)

#### Summary Statistics

Full set of summary statistics (min, max, mean, standard deviation, etc.) for each column of a DataFrame:
```
df.describe()
```

Mean value of each column:
```
df.mean()
```

And similarly for other summary statistics: `df.min()`, `df.max()`, `df.median()`, `df.std()`

Optional keyword argument to `min` and `max` methods, to include only numerical data columns:
```
df.max(numeric_only=True)
```

# Exercise 1

a) Read data for Saskatoon Airport from `'data/weather_yxe.csv'` into a new variable `weather_yxe` and display the first 7 rows.

b) How many rows and columns does `weather_yxe` have? 

c) What are the names and data types of the columns?

d) What are the minimum and maximum relative humidity in this data?

##### Bonus exercises

e) What is the mean wind speed during the first 8 hours (first 8 rows) of data?

f) What are the minimum and maximum relative humidity during the last 10 hours of data?

g) Display summary statistics for all columns for a random sampling of 12 hours of data.

a) Read data for Saskatoon Airport from `'data/weather_yxe.csv'` into a new variable `weather_yxe` and display the first 7 rows. 

In [22]:
weather_yxe = pd.read_csv('data/weather_yxe.csv')
weather_yxe.head(7)

Datetime  Hour of Day     Conditions  Temperature (C)  \
0  2018-05-21 23:00:00           23  Partly Cloudy             17.6   
1  2018-05-22 00:00:00            0   Mainly Clear             16.0   
2  2018-05-22 01:00:00            1   Mainly Clear             13.1   
3  2018-05-22 02:00:00            2   Mainly Clear             12.3   
4  2018-05-22 03:00:00            3   Mainly Clear             12.3   
5  2018-05-22 04:00:00            4   Mainly Clear             12.2   
6  2018-05-22 05:00:00            5  Partly Cloudy             12.1   

   Relative Humidity (%)  Dew Point (C)  Humidex (C)  Pressure (kPa)  \
0                   45.0            6.0          NaN           100.9   
1                   52.0            6.0          NaN           101.0   
2                   61.0            6.0          NaN           101.0   
3                   66.0            6.0          NaN           101.0   
4                   70.0            7.0          NaN           101.1   
5                   76.0            8.0          NaN           101.1   
6                   79.0            9.0          NaN           101.1   

   Visibility (km) Wind Direction  Wind Speed (km/hr)  Wind Chill (C)  
0             24.0             NE                17.0             NaN  
1             24.0            ENE                10.0             NaN  
2             24.0            NNE                11.0             NaN  
3             24.0              N                11.0             NaN  
4             24.0            NNE                16.0             NaN  
5             24.0            NNE                16.0             NaN  
6             24.0            ENE                16.0             NaN

- Compare the data file in the CSV viewer and with the table displayed here 
- `NaN` means "not a number", i.e. missing data

b) How many rows and columns does `weather_yxe` have? 

In [23]:
weather_yxe.shape

(24, 12)

24 rows and 12 columns

c) What are the names and data types of the columns?

In [24]:
weather_yxe.dtypes

Datetime                  object
Hour of Day                int64
Conditions                object
Temperature (C)          float64
Relative Humidity (%)    float64
Dew Point (C)            float64
Humidex (C)              float64
Pressure (kPa)           float64
Visibility (km)          float64
Wind Direction            object
Wind Speed (km/hr)       float64
Wind Chill (C)           float64
dtype: object

d) What are the minimum and maximum wind speeds in this data?

We can use `max` and `min`, or `describe`. Here's the `describe` version:

In [25]:
weather_yxe.describe()

Hour of Day  Temperature (C)  Relative Humidity (%)  Dew Point (C)  \
count    24.000000        24.000000              24.000000      24.000000   
mean     11.500000        19.104167              49.833333       7.291667   
std       7.071068         5.453956              19.771888       1.966661   
min       0.000000        12.100000              23.000000       4.000000   
25%       5.750000        14.150000              31.500000       6.000000   
50%      11.500000        17.700000              48.000000       7.000000   
75%      17.250000        24.550000              69.000000       8.250000   
max      23.000000        26.600000              79.000000      11.000000   

       Humidex (C)  Pressure (kPa)  Visibility (km)  Wind Speed (km/hr)  \
count          1.0       24.000000        24.000000           24.000000   
mean          24.0      101.295833        23.791667           10.541667   
std            NaN        0.192194         1.020621            4.881012   
min           24.0      100.900000        19.000000            0.000000   
25%           24.0      101.100000        24.000000            7.750000   
50%           24.0      101.400000        24.000000           10.500000   
75%           24.0      101.400000        24.000000           14.000000   
max           24.0      101.500000        24.000000           20.000000   

       Wind Chill (C)  
count             0.0  
mean              NaN  
std               NaN  
min               NaN  
25%               NaN  
50%               NaN  
75%               NaN  
max               NaN

The minimum relative humidity is 23% and the maximum is 79%.

e) We can use `describe` or `mean` methods. Here is an example solution using `mean`.

In [26]:
first_eight = weather_yxe.head(8)
first_eight.mean()

Hour of Day                5.500
Temperature (C)           13.850
Relative Humidity (%)     64.750
Dew Point (C)              7.250
Humidex (C)                  NaN
Pressure (kPa)           101.050
Visibility (km)           23.375
Wind Speed (km/hr)        14.625
Wind Chill (C)               NaN
dtype: float64

The mean wind speed in the first 8 hours is 14.625 km/hr.

f) We could create a new variable `last_ten = weather_yxe.tail(10)`, as in the previous example, or we can skip that step and "chain" the methods together:

In [27]:
weather_yxe.tail(10).describe()

Hour of Day  Temperature (C)  Relative Humidity (%)  Dew Point (C)  \
count     10.00000        10.000000              10.000000      10.000000   
mean      17.50000        24.040000              31.700000       6.000000   
std        3.02765         3.320375               8.654479       1.154701   
min       13.00000        16.500000              23.000000       4.000000   
25%       15.25000        23.650000              25.500000       5.250000   
50%       17.50000        25.250000              29.000000       6.000000   
75%       19.75000        26.250000              34.250000       6.750000   
max       22.00000        26.600000              51.000000       8.000000   

       Humidex (C)  Pressure (kPa)  Visibility (km)  Wind Speed (km/hr)  \
count          1.0       10.000000             10.0           10.000000   
mean          24.0      101.410000             24.0            9.700000   
std            NaN        0.056765              0.0            4.498148   
min           24.0      101.300000             24.0            0.000000   
25%           24.0      101.400000             24.0            7.250000   
50%           24.0      101.400000             24.0           11.000000   
75%           24.0      101.400000             24.0           13.000000   
max           24.0      101.500000             24.0           14.000000   

       Wind Chill (C)  
count             0.0  
mean              NaN  
std               NaN  
min               NaN  
25%               NaN  
50%               NaN  
75%               NaN  
max               NaN

The minimum and maximum relative humidity during the last 10 hours are 23% and 51%, respectively.

g) Since the `sample` method gives a different random sampling each time, the numbers will be different in your output compared to the example solution below.

In [28]:
weather_yxe.sample(12).describe()

Hour of Day  Temperature (C)  Relative Humidity (%)  Dew Point (C)  \
count    12.000000        12.000000              12.000000      12.000000   
mean     10.833333        19.433333              49.916667       7.333333   
std       6.617241         6.310645              22.869028       2.188122   
min       1.000000        12.100000              23.000000       4.000000   
25%       4.750000        12.900000              29.500000       6.000000   
50%      11.500000        20.600000              48.000000       7.000000   
75%      15.750000        25.125000              71.500000       8.250000   
max      20.000000        26.600000              79.000000      11.000000   

       Humidex (C)  Pressure (kPa)  Visibility (km)  Wind Speed (km/hr)  \
count          1.0        12.00000             12.0           12.000000   
mean          24.0       101.30000             24.0           11.500000   
std            NaN         0.17581              0.0            3.872983   
min           24.0       101.00000             24.0            4.000000   
25%           24.0       101.10000             24.0            9.500000   
50%           24.0       101.40000             24.0           12.000000   
75%           24.0       101.40000             24.0           14.500000   
max           24.0       101.50000             24.0           16.000000   

       Wind Chill (C)  
count             0.0  
mean              NaN  
std               NaN  
min               NaN  
25%               NaN  
50%               NaN  
75%               NaN  
max               NaN

# Interlude: So What?

- Why bother with any of this? 
- Isn't it easier to do all these tasks in Excel?
- Why would I care about the shape of a DataFrame, or printing out the column names and data types? 
  - Can't I just look at a spreadsheet and this information is obvious, without writing any code?

There are plenty of cases in which Python is overkill and Excel (or other application) is the perfect tool for the job.

In many other cases, for example when your data is large and unwieldy, even simple Python commands can hugely simplify tasks that would be very difficult and time consuming to do within a spreadsheet.

- For an example, let's look at: `Example Notebook - Python Developers Survey`